Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -1625,16 +1653,112 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     fireTabsReplacedEventIfNeeded(oldTab.identifier(), newTab.identifier());
 }
 
-void WebExtensionContext::didChangeTabProperties(const WebExtensionTab& tab, OptionSet<WebExtensionTab::ChangedProperties> properties)
+void WebExtensionContext::didChangeTabProperties(WebExtensionTab& tab, OptionSet<WebExtensionTab::ChangedProperties> properties)
 {
     ASSERT(tab.extensionContext() == this);
 
-    RELEASE_LOG_DEBUG(Extensions, "Changed tab properties (%{public}X) for tab %{public}llu", properties.toRaw(), tab.identifier().toUInt64());
+    RELEASE_LOG_DEBUG(Extensions, "Changed tab properties (0x%{public}X) for tab %{public}llu", properties.toRaw(), tab.identifier().toUInt64());
 
     if (!isLoaded() || !tab.extensionHasAccess())
         return;
 
-    fireTabsUpdatedEventIfNeeded(tab.parameters(), tab.changedParameters(properties));
+    bool hasChangesPending = !tab.changedProperties().isEmpty();
+    tab.addChangedProperties(properties);
+
+    // If there are already changes pending, don't schedule the event to fire again.
+    if (hasChangesPending)
+        return;
+
+    constexpr auto updatedEventDelay = 25_ms;
+
+    // Fire the updated event after a small delay to coalesce relevant changes together.
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(updatedEventDelay.seconds() * NSEC_PER_SEC)), dispatch_get_main_queue(), makeBlockPtr([this, protectedThis = Ref { *this }, tab = Ref { tab }]() {
+        RELEASE_LOG_DEBUG(Extensions, "Firing updated tab properties (0x%{public}X) for tab %{public}llu", tab->changedProperties().toRaw(), tab->identifier().toUInt64());
+        fireTabsUpdatedEventIfNeeded(tab->parameters(), tab->changedParameters());
+        tab->clearChangedProperties();
+    }).get());
+}
+
+void WebExtensionContext::didStartProvisionalLoadForFrame(WebPageProxyIdentifier pageID, WebExtensionFrameIdentifier frameID, WebExtensionFrameIdentifier parentFrameID, const URL& targetURL, WallTime timestamp)
+{
+    auto tab = getTab(pageID);
+
+    // Dispatch webNavigation events.
+    if (tab && hasPermission(_WKWebExtensionPermissionWebNavigation) && hasPermission(targetURL)) {
```

Comment:
Do we want to pass tab to `hasPermission` so activeTab checks will happen? Or does that happen even without `tab` being passed?

Reply:
Passing `tab` for the feature `hasPermission` is only needed for `tabs` permission, but I might as well do it here just for coverage in case that ever changes!

Reply:
Oh, but you mean `hasPermission` for the URL! Yes, that is needed to work with `activeTab`.